Cleanup/overhaul of DSDHacked support

electricbrass opened the pull request:
This is an attempt to fix many of the remaining bugs with the DSDHacked implemenation as well as clean up the code a bit.

It includes fixes for the following:
- Legacy of Rust lamps playing "misc/teamchat" when broken
- Weapon sprites in certain index ranges not rendering
- Modified vanilla things only using their dehacked modifications when not directly placed in the map
- A few bad format strings
- Avatars not being spawned on clients in multiplayer
- CTF things not being spawned on clients in multiplayer
- Missing GPL headers on new files
- SNDINFO ambient sound things not spawning

As well as the following features:
- Support for ID24 `USER_` and `DEHTHING_` strings (this allows using DSDHacked things in HORDEF!)
- Support for ID24's `bossactionednum` UMAPINFO key

electricbrass committed:
Start rewriting DoomObjectContainer so that it owns the objects instead of storing pointers

electricbrass committed:
It compiles now, still broken

electricbrass committed:
Further progress towards cleaning up dsdhacked memory allocation. Eviternity II things are missing though

electricbrass committed:
Fix patchthing and patchframe

electricbrass committed:
Remove some more unnecessary allocations

electricbrass committed:
Restore original Z_Init location

electricbrass committed:
Fix rangecheck in R_DrawPSprite and update it from C89

electricbrass committed:
Clean up use of sprnames size and fix P_SpawnMapThing

electricbrass committed:
Fix spawn map containing separate copies of mobjinfo_ts

electricbrass committed:
Change DoomObjectContainer::operator[] to return a reference instead of a pointer

electricbrass committed:
Use int32_t in a couple more spots instead of mobjtype_t

electricbrass committed:
Fix server build

electricbrass committed:
Simplify deh backup

electricbrass committed:
Default constructor and destructor for doomobjectcontainer

electricbrass committed:
Remove unnecessary typedefs

electricbrass committed:
Remove useless M_Free_Ref

electricbrass committed:
Remove faulty z_realloc

It's a useful idea. If we need it, we can add it later with a proper implementation

electricbrass committed:
More minor cleanup

electricbrass committed:
More cleanup and improved error message for sprites

electricbrass committed:
Restore the in-order table to doomobjcontainer

electricbrass committed:
Fix teamchat sound bug

electricbrass committed:
Clean up DoomObjectContainer by moving all methods into the class since its header only anyway and make its much less verbose

electricbrass committed:
Fix compilation errors

electricbrass committed:
Move spawn map dehacked updates to be performed when setting the editor number

cmbernard333 commented on common/doom_obj_container.h:
```diff
@@ -1,315 +1,199 @@
-	return this->find(idx) != this->end();
-}
+	// TODO: do we ever need this?
+	void append(const DoomObjectContainerType& other)
```
Shouldn't need it unless the new `insert` doesn't support appending another doom object container. Backup is the only use case I can think of something like this.

electricbrass replied:
insert doesn't support that but even for backup, we don't really ever want to be appending. I think it's probably safe to just remove this entirely.

cmbernard333 commented on common/doom_obj_container.h:
```diff
@@ -1,315 +1,199 @@
-{
-	for (auto& obj : this->container)
+	using LookupTable = std::unordered_map<IdxType, ObjType*>;
+	// TODO: maybe use the vector for iteration instead of the map for speed?
```
Depends on the type of iteration. The in order table has _everything_ that has been added via initialization and dehacked. The map should always contain the working set so iterating over that is usually the most accurate of what objects are used in game.

electricbrass committed:
Rename DoomObjectContainer file for consistency

electricbrass committed:
Implment DoomObjectContainer::rebuildMap for proper sorting of populated containers

electricbrass committed:
Remove state.h/.cpp as its contents can now be condensed into 3 lines

electricbrass committed:
Remove D_Init_Nightmare_Flags

Those flags are already set inside the mobjinfo array

electricbrass committed:
Add default values for mobjinfo_t fields

electricbrass committed:
Finish moving fully over to new doomobjectcontainer interface

electricbrass committed:
Fix negative index thing spawns not being communicated to client

electricbrass committed:
Remove sprite.h/.cpp

The only function left in it was never used

electricbrass committed:
Fix up orgsprnames

electricbrass committed:
Fix clientside weapon state parsing

electricbrass committed:
Just load from backups instead of reiniting the state/mobjinfo/sprnames lookup tables

electricbrass committed:
Dehacked cleanup

electricbrass committed:
Remove unneeded DoomObjectContainer::append

electricbrass committed:
Merge branch 'protobreak' into dsdhacked-rewrite

electricbrass committed:
If a specific type is specified, search mobjinfo, not the spawn map in P_SpawnMapThing.

This matches pre-DSDHacked behavior and fixes spawning of things such as ambient sound sources

electricbrass committed:
Implement bossactionednum

electricbrass committed:
Fix not being able to patch mbf pointers

electricbrass committed:
Add support for id24 USER_ and DEHTHING_ strings

electricbrass committed:
Bump infinite state cycle detection limit to 128 (SD21 has a cycle greater than 64).

Also fix the error message to not crash when an unnamed mobj causes a cycle

electricbrass committed:
Merge branch 'protobreak' into dsdhacked-rewrite
